### ISSUE TITLE
Implement whisper.cpp

### ISSUE THREAD

**Issue by ericcurtin** [containers/ramalama#51]
If there is a way to auto-detect between language model files and asr model files. We should do that, we should just use a runtime flag, so some options would be:

```
ramalama --runtime vllm
ramalama --runtime llama.cpp
ramalama --runtime whisper.cpp
```

**Comment by rhatdan** [comment]
Can this be closed?  Do we have this functionality now?

**Comment by ericcurtin** [comment]
There's still more work for both `--runtime whisper.cpp` and `--runtime vllm` ... It can be tracked under this issue or somewhere else.

**Comment by ericcurtin** [comment]
I would say just this to start:

`ramalama --runtime run whisper.cpp ggml-large-v3-turbo.bin`

which would just perform a fairly standard whisper.cpp command.

No interactive support, it's not the same as a chat bot workflow.

We have renovate controlling what version of whisper.cpp we build against and it runs everything through CI before we rebase, I'd like to keep this, at least there's a CI run before we suddenly change version. Just cloning main/master without CI, I'd rather not.

**Comment by ericcurtin** [comment]
@p5 kindly set up renovate for us.

**Comment by FNGarvin** [comment]
> `ramalama --runtime run whisper.cpp ggml-large-v3-turbo.bin`
> 
> which would just perform a fairly standard whisper.cpp command.

Could you elaborate, please?  What command would be performed?  Would that be provided through additional command-line arguments to the command you've just given?  Would you pass that as though it were a prompt to another model?

The boilerplate syntax for using whisper.cpp directly is something like ```./main -m /models/ggml-large-v3-turbo.bin -f /audios/jfk.wav```  Could you please give me an example of the kind of command-line you're envisioning to complete the same task via Ramalama?  

> No interactive support, it's not the same as a chat bot workflow.

Sorry to be dense, but I can't tell if you're reiterating that whisper.cpp does not provide interactive support or if you're saying that you do not like the concept of dropping the user to a shell prompt with some workspace-like features.

**Comment by ericcurtin** [comment]
I meant something like this, corrected:

`ramalama --runtime whisper.cpp run ggml-large-v3-turbo.bin jfk.wav`

**Comment by ericcurtin** [comment]
> Sorry to be dense, but I can't tell if you're reiterating that whisper.cpp does not provide interactive support or if you're saying that you do not like the concept of dropping the user to a shell prompt with some workspace-like features.

We are always open to ideas. It's just less relevant with whisper.cpp, one doesn't speak to an interactive prompt. But if someone finds that useful for some reason, always happy to look at PRs, etc.

**Comment by ericcurtin** [comment]
This PR is a perfect example of why we don't just clone main/master:

https://github.com/containers/ramalama/pull/474

**Comment by ericcurtin** [comment]
> Yes, failing each time whisper.cpp is updated is perhaps a better solution than blindly updating along with it. Anyway, back to the meat of my question...
> 
> > It's just less relevant with whisper.cpp, one doesn't speak to an interactive prompt. But if someone finds that useful for some reason, always happy to look at PRs, etc.
> 
> I don't care at all about that. I just saw the seemingly abandoned whisper.cpp stub in the container and thought to make it usable. It is not currently usable, right? Do you care to see it made usable?

Yup we sure do, this issue is open for someone to complete it :)

**Comment by FNGarvin** [comment]
> this issue is open for someone to complete it :)

Great.  I'm just having trouble understanding what "complete it" means to you.  That's why I'm asking about the example command-lines you're envisioning.  

> The boilerplate syntax for using whisper.cpp directly is something like ./main -m /models/ggml-large-v3-turbo.bin -f /audios/jfk.wav Could you please give me an example of the kind of command-line you're envisioning to complete the same task via Ramalama?

**Comment by ericcurtin** [comment]
> > this issue is open for someone to complete it :)
> 
> Great. I'm just having trouble understanding what "complete it" means to you. That's why I'm asking about the example command-lines you're envisioning.

We are open to ideas. Rome wasn't built in a day, one PR at a time.

> 
> > The boilerplate syntax for using whisper.cpp directly is something like ./main -m /models/ggml-large-v3-turbo.bin -f /audios/jfk.wav Could you please give me an example of the kind of command-line you're envisioning to complete the same task via Ramalama?

Getting this executing the main command you specified would be a start

ramalama --runtime whisper.cpp run ggml-large-v3-turbo.bin jfk.wav

**Comment by FNGarvin** [comment]
Thanks.  And in the case that Ramalama is running the inferencing in a container, how we get the input media into the model?  Would you infer a directory to bind based on the working directory of the file given as an argument?

**Comment by rhatdan** [comment]
Yes the wav file would need to be volume mounted into the container with a :z option.

Would it make sense to also allow stdin for a way file?

cat jfk.wav | ramalama --runtime whisper.cpp run ggml-large-v3-turbo.bin -

I think it makes sense to allow grab output from stdout.

cat jfk.wav | ramalama --runtime whisper.cpp run ggml-large-v3-turbo.bin - > /tmp/output

Not sure what ramalama serve would do?

**Comment by FNGarvin** [comment]
Thank you for helping to bring me up to speed with some mock usage examples.  I'm still pretty novice at containers, but could we bind the directory read-only instead of using :z?  A dryrun (*love* that feature, btw) of something like(?):

        podman run --rm -i --device nvidia.com/gpu=all \
        --mount=type=image,src=WHISPER.CPP-MODEL-SHORTNAME,destination=/mnt/models,rw=false,subpath=/models \
        --mount=type=bind,src=/[FQual-dir-of-user-inputfilewav]/,destination=/mnt/audio,rw=false \
        ramalama-image /bin/sh -c "whisper-main -m /mnt/models/model.file -f /mnt/audio/USERFILE.wav"

> Would it make sense to also allow stdin for a way file?

It's more appealing in principle than binding directories, but whether or not it is practical is currently unknown to me.  It has been a verrrrry long time since I've relied on a shell to create pipes for large data and I don't quite remember what the gotchas were, but I'm thinking there were at least a few.  I know even less about what might happen trying to pipe potentially massive, uncompressed audio files into the container on STDIN.  

> I think it makes sense to allow grab output from stdout.

Redirects aren't adequate?

> Not sure what ramalama serve would do?

If the documentation on the whisper.cpp github is accurate, it seems like the provided web service is very basic.  All the examples are making requests via curl from the command-line.  I totally get it and that isn't criticism, but we're not talking about a convenient UI AFAICT.  I am aware that there exist some third-party front-ends, like https://github.com/litongjava/whisper-cpp-server + https://github.com/litongjava/listen-know-web, but I don't really know anything about them.

For me, personally, the most likely use-case for whisper.cpp is in generating subtitles for arbitrary videos (upon audio extracted w/ ffmpeg) or for transcribing and translating arbitrary audio sequences.  Command-line invocation over a bound directory seems adequate and possibly ideal.  It probably doesn't require any code overlap with whisper.cpp, using Ramalama as scaffolding to bring all the pieces together and offering a layer of abstraction wrt GPU config, system libraries, etc.  But I'm not deep into any of these techs or projects, so if there's a better vision / direction I'd love to hear about it.

**Comment by rhatdan** [comment]
The stdout stuff should just work, we could even just have whisper grab /dev/stdin when it sees the "-". 

The volume mount can also be marked ro,z so it is both readonly and available for reading based on SELinux.

**Comment by rhatdan** [comment]
podman run --rm -i --device nvidia.com/gpu=all \
    --mount=type=image,src=WHISPER.CPP-MODEL-SHORTNAME,destination=/mnt/models,rw=false,subpath=/models \
    --mount=type=bind,src=/[FQual-dir-of-user-inputfilewav]/,destination=/mnt/audio/`USERFILE.wav`,rw=false,z \
    ramalama-image /bin/sh -c "whisper-main -m /mnt/models/model.file -f /mnt/audio/USERFILE.wav"

**Comment by rhatdan** [comment]
Slightly simpler.

podman run --rm -i --device nvidia.com/gpu=all \
--mount=type=image,src=WHISPER.CPP-MODEL-SHORTNAME,destination=/mnt/models,rw=false,subpath=/models \
-v/[FQual-dir-of-user-inputfilewav]/:/mnt/audio/USERFILE.wav:ro,z \
ramalama-image /bin/sh -c "whisper-main -m /mnt/models/model.file -f /mnt/audio/USERFILE.wav"

**Comment by rhatdan** [comment]
Looks like `whisper -f -` is supported now, although the entire file needs to be flushed through the pipe I believe.

**Comment by FNGarvin** [comment]
> Looks like whisper -f - is supported

It does look that way, though I haven't prepared any large wav files to test with.  Thank you - your way is much better, I think, than binding a directory for a process that will only require one input.  Especially if it creates the possibility of chaining ffmpeg without intermediate conversion files.

So, consensus seems to be that ```cat jfk.wav | ramalama --runtime whisper.cpp run ggml-tiny.bin``` should produce and run something vaguely like ```podman run --rm -i --device nvidia.com/gpu=all --mount=type=bind,src=models/,destination=/mnt/models,rw=false ramalama /bin/sh -c "whisper-main -m /mnt/models/ggml-tiny.bin -f -"```  That doesn't seem too difficult at first blush.  I'll see what I can do in the coming days.

Thanks

**Comment by ericcurtin** [comment]
stdin seems fine to me. Also we can autodetect when there is stdin coming in, it's better for usability, you don't need the explicit '-' then, although we can have the ability to explicitly use stdin also '-', grep is an example of a command that does this and llama-run. Once llama-run is integrated into RamaLama, I plan on adding this to the ramalama run:

`git diff | ramalama run granite-code "Write a git commit message for this change"`

llm and ollama have this feature.

**Comment by rhatdan** [comment]
Cool

**Comment by rhatdan** [comment]
We now have images with whisper-server as an entrypoint.  Should be able to run one of these images directly and redirect the WAVE file to it.

**Comment by rhatdan** [comment]
@ericcurtin not sure what the work needs to be done for this one?

**Comment by rhatdan** [comment]
@ericcurtin What should we do with this one now?

**Comment by FNGarvin** [comment]
Hi all, just wanted to add a quick follow-up to this thread since I participated in the discussion.

At the time, my primary use case was generating subtitle files (e.g., SRT) for a side project. After our discussion, it became clear that the most direct path for Ramalama was to support piping `whisper.cpp`'s standard output, while my need was more centered around its file-based output features (like `-osrt`).

Since my project priorities were tied to that specific workflow, I ended up moving in a different direction and didn't have the bandwidth to circle back to implement the `stdin`/`stdout` plan we discussed. Apologies for not closing the loop sooner.

It's great to see the issue is still active and that you're still actively improving Ramalama.  It's a great and undervalued tool.

-FNG

**Comment by aenertia** [comment]
Running whisper through ramalama to manage using correct cuda/rocm environment would be very useful not just for whisper (so you could i.e interact with it like 'normal' i.e ```whisper video.mp4 --model base --output_format srt --language en``` would do what you'd expect via ramalama run "whisper video.mp4 --model base --output_format srt --language en" or something similar. I think whilst you can basically cobble this with direct podman pokes. Having some support for 'one-shots' for a range of models is probably useful. Maybe likewise for implementing llama-swap like functionality likewise.

**Comment by FNGarvin** [comment]
I don't know if it's useful information, but I can share my findings in working through getting a containerized whisper.cpp subtitle task going.  

Firstly, building the image kind of sucks.  Mad, mad respect to the project and everything else Mr. Gerganov and co are doing, but there aren't exactly a ton of usable binary releases.  So even though it isn't the most pattern-correct strategy, it kind of makes sense to package llama.cpp and whisper.cpp on a single image.  Or at least different stages of a common base.  But in working through the subtitle task, I came to find that working with whisper.cpp required a very different workflow than working with llama.cpp.

Juggling the srt file (that evidently can't be directed to stdout) is only one hurdle preventing light-touch shuttling of parameters and IO back and forth for srt generation.  In my testing using cribbed Youtube videos selected for diversity of content in a variety of languages and a variety of subtitle densities, a shocking number of files had sufficient corruption to require using ```-err_detect ignore_err``` just to get through a ffprobe or ffmpeg pass.  Also, the ONLY way I could reliably get quality transcriptions with a minimum of hallucination was to use the LARGE model and to perform noise gating.  FFMPEG can do this but, between the noise gating and the ignored encode errors, the prepared wav file almost always had a meaningfully different duration than the source video.  Programmatically detecting the length of silence at the video start (and possibly end) combined with timeshifting all the timestamps in-between was possible and generally satisfactory... but you would frequently want/need to inspect the processing artifacts (the temporary wav(s), the pre-timestretched srt, etc) to understand and fine-tune the process for any particularly tough inputs.  

Though I am hardly qualified to gauge, this to me no longer seems like Ramalama modus operand-ai (haha).  At the present levels of the tech, I don't see any way the process can be simplified to let Ramalama enable the srt workflow in a way of operation that has transparent semantics similar to everything else it does.  But I could be wrong!  I am not a Python coder and can't claim that this is good, problem-free, or useful but my wrapper ended up looking something like this: https://gist.github.com/FNGarvin/90e48055a61bf1fd08f9e2625e427226

**Comment by rhatdan** [comment]
This is not stale.

**Comment by Tvpower** [comment]
Hi my name is Mario Pinzon. I'm trying to participate in Hacktoberfest and do more open source contributions is there any way this task could be assigned to me? I've been reading this repo for some time and I would love if I could help :D

Lmk!

**Comment by Tvpower** [comment]
From my understanding, the core requirement was to implement:
`ramalama --runtime whisper.cpp run model.bin audio.wav`

Based on the discussion, I believe the main work involves:
1. Adding runtime detection logic for whisper.cpp
2. Implementing command translation to bridge RamaLama's interface with whisper.cpp's native commands
3. Proper volume mounting for audio files
4. stdin/stdout handling for piped workflows

I was reading @FNGarvin earlier implementation which I saw that it focused more on subtitles generation with ffmpeg processing.

If you dont mind me asking @ericcurtin or @rhatdan confirm if this is correct? Are there specific acceptance criteria or test I should target? Any guidance is extremely appreciated!

**Comment by rhatdan** [comment]
Yes we want to be able to do something like that but also

cat audio.wav | ramalama --runtime whisper.cpp run model.bin

**Comment by rhatdan** [comment]
Assigned

**Comment by tulilirockz** [comment]
I started working on this because it seemed fun, I have a PR up here: https://github.com/containers/ramalama/pull/2038

**Comment by rhatdan** [comment]
This is still an open issue if others are interested in working on it, or following up on @tulilirockz PR.
